core: add v2 conflict detection endpoint

Erashin opened the pull request:
Fixes #6905. Convert V2 payload to V1 structs to use them in `Conflicts.kt`, because stdcm is not v2 yet and still calls conflict detection with v1 structs.

Erashin committed:
core: add v2 conflict detection endpoint

eckter commented on core/src/main/kotlin/fr/sncf/osrd/api/api_v2/CommonSchemas.kt:
```diff
@@ -30,3 +31,39 @@ class SignalSighting(
+    @Json(name = "begin_time") val beginTime: TimeDelta,
+    @Json(name = "end_time") val endTime: TimeDelta,
+)
+
+/**
+ * Dates can be typed as `LocalDateTime` in json compatible types, with this adapter it will be
+ * converted to/from ISO8601 strings
+ */
+class DateAdapter : JsonAdapter<ZonedDateTime>() {
+    @FromJson
+    override fun fromJson(reader: JsonReader): ZonedDateTime {
+        return ZonedDateTime.parse(reader.nextString())
+    }
+
+    @ToJson
+    override fun toJson(writer: JsonWriter, value: ZonedDateTime?) {
+        writer.value(value?.toString())
+    }
+}
```
It should be moved to the file `UnitJsonAdapters`, and can be integrated to `UnitAdapterFactory` so that it doesn't need to be added to each adapter

Erashin replied:
I didn't put it there as it is not really a unit, as Distance, Duration etc. I don't mind putting it there though.

eckter commented on core/src/main/kotlin/fr/sncf/osrd/api/api_v2/conflicts/ConflictDetectionEndpointV2.kt:
```diff
@@ -0,0 +1,95 @@
+): List<TrainRequirements> {
+    val trainRequirements = mutableListOf<TrainRequirements>()
+    for ((id, trainReq) in trainReqs) {
+        val delta = TimeDelta(between(trainReq.startTime, startTime).toMillis())
```
I suspect that the math is off: `between(start, end)` returns a negative value if `end < start`. Considering `startTime` is the earliest value, `delta` would be `<=0`. Which is not necessarily a problem in itself, but I don't think *adding* that negative delta would offset the times in the correct way.

eckter commented on core/src/main/kotlin/fr/sncf/osrd/api/api_v2/conflicts/ConflictDetectionEndpointV2.kt:
```diff
@@ -0,0 +1,95 @@
+        conflicts.map {
+            Conflict(
+                it.trainIds,
+                startTime.plusSeconds(it.startTime.toLong()),
+                startTime.plusSeconds(it.startTime.toLong()),
```
We lose some precision here, the seconds are rounded down. We could instead do something like `startTime.plus(Duration.ofMillis((it.startTime * 1000).toLong()))`

Though that is quite ugly. 

By the way the second line should be changed to `endTime`

eckter commented on core/src/main/kotlin/fr/sncf/osrd/api/api_v2/conflicts/ConflictDetectionResponse.kt:
```diff
@@ -0,0 +1,28 @@
+    @Json(name = "start_time") val startTime: ZonedDateTime,
+    @Json(name = "end_time") val endTime: ZonedDateTime,
+    @Json(name = "conflict_type")
+    val conflictType: ConflictDetectionEndpoint.ConflictDetectionResult.Conflict.ConflictType,
```
You could add an import there

eckter commented on core/src/main/kotlin/fr/sncf/osrd/api/api_v2/conflicts/ConflictDetectionResponse.kt:
```diff
@@ -0,0 +1,28 @@
+)
+
+val conflictResponseAdapter: JsonAdapter<ConflictDetectionResponse> =
+    Moshi.Builder()
+        .addLast(DateAdapter())
+        .addLast(KotlinJsonAdapterFactory())
+        .build()
+        .adapter(ConflictDetectionResponse::class.java)
```
I haven't tested, but I don't think the distances will be handled properly (missing unit adapter)

Erashin replied:
There are no distances/time deltas here, so the unit adapter is not necessary. However, it will become so when i put the date adapter in it :)

eckter commented on core/kt-osrd-utils/src/main/kotlin/fr/sncf/osrd/utils/json/UnitJsonAdapters.kt:
```diff
@@ -67,12 +68,29 @@ class DurationAdapter : JsonAdapter<Duration?>() {
 }
 
+/**
+ * Dates can be typed as `LocalDateTime` in json compatible types, with this adapter it will be
```
```suggestion
 * Dates can be typed as `ZonedDateTime` in json compatible types, with this adapter it will be
```

woshilapin commented on editoast/src/core/v2/simulation.rs:
```diff
@@ -224,13 +227,30 @@ pub struct SimulationRequest {
+// Since there is only on success and others are error cases
+#[derive(Serialize, Deserialize, Clone, Debug, ToSchema)]
+#[serde(tag = "status", rename_all = "snake_case")]
+#[allow(clippy::large_enum_variant)]
```
I believe I remember why we do allow in this case (success variant is the biggest one, and it's "ok-ish" to waste memory space on error path). But in general, I would expect a comment above every `#[allow(...)]` to explain why we do allow this exception.

woshilapin replied:
Haha, it was already there, just a few lines above 🤦

woshilapin commented on editoast/src/views/v2/timetable.rs:
```diff
@@ -246,40 +257,104 @@ async fn delete(
+            train_names: c
+                .train_ids
+                .iter()
+                .map(|id| trains.get(id).unwrap().train_name.clone())
```
Here's a `.unwrap()` that I believe should be removed one way or another. From what I understand, the `.unwrap()` assumes that `core` and `editoast` has the same data loaded (what if some data has been changed in `editoast` in-between the core request?). Also, it shows also the problem of data redundancy in `Conflict` (we have `CoreConflit` in it, which already contains the trains, and we also store the same trains in another field).

I don't have a proposition here, but I'll be happy to discuss it to try to find one together.

woshilapin commented on editoast/src/core/v2/conflict_detection.rs:
```diff
@@ -0,0 +1,59 @@
+
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, ToSchema)]
+pub enum ConflictType {
+    Spacing,
+    Routing,
```
I'm wondering if it would be useful to actually document a bit more what 'Spacing' conflict or 'Routing' conflict means. Because 3 weeks ago, I would have found this code and would have had no idea what this means. Since this is part of the OpenAPI, documentation is probably a good idea?

woshilapin reviewed (changes requested):
I've reviewed only the `editoast` part. There is at least one comment about the redundancy of information that I'd like to be addressed (or be convinced it's good as-is). The 2 other comments are minor. Apart from that, thanks for the good work, and long live v2!

woshilapin reviewed (approved):
Thanks for the pair programming, looks good for `editoast` 🎉

Erashin committed:
core: handle failing pathfinding and simulation

Co-authored-by: Florian Amsallem <florian.amsallem@gmail.com>

flomonster committed:
editoast: add call to conflict detection core endpoint

- Handle batch simulation for better performances
- Rename signal projection endpoint url

woshilapin commented on core/src/main/java/fr/sncf/osrd/cli/ApiServerCommand.java:
```diff
@@ -95,8 +96,9 @@ public int run() {
-                    new FkRegex("/v2/project_signals", new SignalProjectionEndpointV2(infraManager)),
+                    new FkRegex("/v2/signal_projection", new SignalProjectionEndpointV2(infraManager)),
                     new FkRegex("/detect_conflicts", new ConflictDetectionEndpoint()),
+                    new FkRegex("/v2/conflict_detection", new ConflictDetectionEndpointV2()),
```
Maybe these endpoints should be added to the `core/openapi.yaml`.
